include/alpaka/mem/view/Accessor.hpp

```diff
@@ -0,0 +1,400 @@
+ */
+
+#pragma once
+
+#include <alpaka/dim/Traits.hpp>
+#include <alpaka/mem/view/Traits.hpp>
+
+namespace alpaka
+{
+    namespace internal
+    {
+        template<typename T>
+        auto asBytePtr(T* p)
+        {
+            return reinterpret_cast<char*>(p);
+        }
+        template<typename T>
+        auto asBytePtr(const T* p)
+        {
+            return reinterpret_cast<const char*>(p);
+        }
+    } // namespace internal
+
+    struct ReadAccess
+    {
+    };
+    struct WriteAccess
+    {
+    };
+    struct ReadWriteAccess
+    {
+    };
+
+    template<typename MemoryHandle, typename Elem, typename BufferIdx, std::size_t Dim, typename AccessModes>
+    struct Accessor;
+
+    namespace internal
+    {
+        template<typename T>
+        struct WriteOnlyProxy
+        {
+            WriteOnlyProxy(T& location) : loc(location)
+            {
+            }
+
+            template<typename U>
+            auto& operator=(U&& value)
+            {
+                loc = std::forward<U>(value);
+                return *this;
+            }
+
+        private:
+            T& loc;
+        };
+
+        template<typename Pointer, typename Elem, typename AccessModes>
+        struct AccessReturnTypeImpl;
+
+        template<typename Pointer, typename Elem>
+        struct AccessReturnTypeImpl<Pointer, Elem, ReadAccess>
+        {
+            using type = Elem;
+        };
+
+        template<typename Pointer, typename Elem>
+        struct AccessReturnTypeImpl<Pointer, Elem, WriteAccess>
+        {
+            using type = WriteOnlyProxy<Elem>;
+        };
+
+        template<typename Pointer, typename Elem>
+        struct AccessReturnTypeImpl<Pointer, Elem, ReadWriteAccess>
+        {
+            using type = Elem&;
+        };
+
+        template<typename Pointer, typename Elem, typename HeadAccessMode, typename... TailAccessModes>
+        struct AccessReturnTypeImpl<Pointer, Elem, std::tuple<HeadAccessMode, TailAccessModes...>>
+            : AccessReturnTypeImpl<Pointer, Elem, HeadAccessMode>
+        {
+        };
+    } // namespace internal
+
+    template<typename Pointer, typename Elem, typename AccessModes>
+    using AccessReturnType = typename internal::AccessReturnTypeImpl<Pointer, Elem, AccessModes>::type;
+
+    template<typename Pointer, typename Elem, typename BufferIdx, typename AccessModes>
+    struct Accessor<Pointer, Elem, BufferIdx, 1, AccessModes>
+    {
+        using ReturnType = AccessReturnType<Pointer, Elem, AccessModes>;
+
+        ALPAKA_FN_ACC Accessor(Pointer p_, Vec<DimInt<1>, BufferIdx> extents_) : p(p_), extents(extents_)
+        {
+        }
+
+        template<typename OtherAccessModes>
+        ALPAKA_FN_ACC Accessor(const Accessor<Pointer, Elem, BufferIdx, 1, OtherAccessModes>& other)
+            : p(other.p)
+            , extents(other.extents)
+        {
+        }
+
+        ALPAKA_FN_ACC auto operator[](Vec<DimInt<1>, BufferIdx> i) const -> ReturnType
```

Comment:
This could be rewritten and would save to write a specialisation for 1 to N-dimensional access

```C++
template< size_t Tdim>
ALPAKA_FN_ACC auto operator[](Vec<DimInt<Tdim>, BufferIdx> i) const
{
    static_assert(Dim = TDim, "vector with wrong dimensionallity used to access the data");
    return access(i, std::make_index_sequence<Tdim>);
}

private: 

template< typename TVec std::size_t... TIndices>
ALPAKA_FN_ACC auto access(TVec i, std::index_sequence<TIndices...>) const -> ReturnType
{
    return access(i[TIndices]...);
}
```

Reply:
Sure, but then I need a dimensionality agnostic base class.

Reply:
I just tried to write that. Lifting that one function into a CRTP base would result in a lot more code than I currently heave. I would like to avoid that.

Reply:
OK in that case we should keep the explicit implementations. Thanks for testing.